### ISSUE TITLE
Integrate Shopify Analytics

### ISSUE THREAD

**Issue by rylanharper** [rylanharper/nitrogen#7]
### Goal
Add basic Shopify Analytics to the project, specifically for tracking **page views** and **cart events**.

### Research Notes
1. Shopify uses **Monorail** as their analytics provider.
   - Reference: `hydrogen-react` [`analytics.ts`](https://github.com/Shopify/hydrogen/blob/2a978cd213608ed47938fab8f3695cbf869b371d/packages/hydrogen-react/src/analytics.ts)
   - This provides insight into how Shopify's analytics API works.
2. Found an interesting [article](https://karvedigital.com/insights/integrating-shopify-analytics-into-your-next-js-custom-storefront) on integrating Shopify Analytics in a Next.js environment.

### Implementation Idea
Add a new server proxy file, `shopify-analytics.ts`, to handle the Shopify Analytics API connection. This could allow the frontend to send event data to the backend, which forwards it to Shopify's Monorail endpoint.

Here’s a draft for the proxy handler:

```ts
import type { H3Event } from 'h3';

/**
 * Handles requests to the Shopify Analytics API through an API proxy.
 * @param event - The H3 event containing the request data
 * @returns The response from the Shopify Analytics endpoint
 */
export default defineEventHandler(async (event: H3Event) => {
  const { shopify: options } = useRuntimeConfig();
  const { events } = await readBody(event);

  return await $fetch(
    `${options.storefront}/.well-known/shopify/monorail/unstable/produce_batch`,
    {
      method: 'POST',
      headers: {
        'Content-Type': 'text/plain'
      },
      body: JSON.stringify({
        events,
        metadata: {
          event_sent_at_ms: Date.now()
        }
      })
    }
  );
});

**Comment by obitblack** [comment]
I know you closed this but wondering if you are still looking for help to implement?

**Comment by rylanharper** [comment]
Hey @obitblack 👋

Yes, it would be great to have help integrating this feature! I know it is possible, but I just didn't have the time to fully research how to integrate analytics within this template due to my work schedule these past few months. I'm actually not even 100% sure if the analytics provider (Monorail) mentioned above is correct - I reached out to Shopify and also posted on the Shopify developer forums, but received no answer back.

If you have a solid concept for how to implement this feature, I'd be open to reviewing a PR. It would also be great to hear any insights you may have on this as well so that I can be informed and offer some critiques/guidance.

If you want to take on this task, I can reopen this issue. The Shopify analytics integration is really the only thing separating Nitrogen from having all features found in Hydrogen.

**Comment by obitblack** [comment]
Yes, you can reopen it.  I need to implement it for our retail store.  I think it also has weight on recommendations and other things so it's kind of vital not just for analytics.

I'll keep this issue updated as I figure stuff out.

Thanks,

Jon.
https://obit.black <https://obit.blacka/>

**Comment by rylanharper** [comment]
Sounds good! Appreciate the help on this. I also know a dev who has some insight on integrating analytics in a Nuxt environment, so I'll reach out to him when he is free.

**Comment by plcdnl** [comment]
Any news on this?

**Comment by obitblack** [comment]
still working on itSent from my iPhoneOn Apr 14, 2025, at 4:06 PM, plcdnl ***@***.***> wrote:﻿
  Any news on this?—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>



plcdnl left a comment (rylanharper/nitrogen#7)
Any news on this?

**Comment by r1me75** [comment]
Hi @rylanharper Amazing Repo man! However, did you ever think of creating a Nitrogen Discord Server? I personally would love to join and help too. 

@obitblack Any updates bro?

**Comment by rylanharper** [comment]
@r1me75 No plans for a Discord atm, but possibly in the future. I get emails weekly about this template so I know people are interested, but for now I am updating the this project when I can. For now, you can suggest new features/enhancements as issues or if you have questions about some aspects of the code feel free to post in the discussions. 

Regarding the analytics, I'll take a look into this issue at the end of the month (2-3 weeks).. I've just been swamped with a large-scale ecom site for a client currently, which is wrapping up soon.

**Comment by rylanharper** [comment]
Closing this thread. Will track in issue #23.
